kernel/xpmem_misc.c

```diff
@@ -315,21 +320,138 @@ xpmem_debug_printk_procfs_open(struct inode *inode, struct file *file)
 	return single_open(file, xpmem_debug_printk_procfs_show, NULL);
 }
 
+static long
+user_atol(const char __user *buffer, size_t count)
+{
+	char buf[8];
+	size_t i;
+	unsigned long value = 0, old_value;
+
+	count = min(count, sizeof(buf));
+	if (copy_from_user(buf, buffer, count)) {
+		return -EFAULT;
+	}
+	for (i = 0; i < sizeof(buf); i++) {
+		if (buf[i] > '9' || buf[i] < '0') {
+			break;
+		}
+		old_value = value;
+		value *= 10;
+		value += buf[i] - '0';
+
+		if (value < old_value) {
+			return -EINVAL;
+		}
+	}
+	return value;
+}
+
+static ssize_t
+xpmem_config_max_page_fault_after_write(struct file *file,
+					const char __user *buffer,
+					size_t count, loff_t *ppos)
+{
+	ssize_t ret = count;
+
+	long value = user_atol(buffer, count);
+	if (value < 0) {
+		return -EINVAL;
+	}
+
+	write_lock(&xpmem_config.lock);
+	if (xpmem_config.max_page_fault_before + value >= XPMEM_MAX_PAGE_FAULTS) {
+		ret = -EINVAL;
+	} else {
+		xpmem_config.max_page_fault_after = value;
+	}
+	write_unlock(&xpmem_config.lock);
+	return ret;
+}
+
+static int
+xpmem_config_max_page_fault_after_show(struct seq_file *seq, void *offset)
+{
+	seq_printf(seq, "%lu\n", xpmem_config.max_page_fault_after);
+	return 0;
+}
+
+static int
+xpmem_config_max_page_fault_after_open(struct inode *inode, struct file *file)
+{
+	return single_open(file, xpmem_config_max_page_fault_after_show, NULL);
+}
+
+static ssize_t
+xpmem_config_max_page_fault_before_write(struct file *file,
+					 const char __user *buffer,
+					 size_t count, loff_t *ppos)
+{
+	ssize_t ret = count;
+
+	long value = user_atol(buffer, count);
+	if (value < 0) {
+		return -EINVAL;
+	}
+
+	write_lock(&xpmem_config.lock);
+	if (xpmem_config.max_page_fault_after + value >= XPMEM_MAX_PAGE_FAULTS) {
+		ret = -EINVAL;
+	} else {
+		xpmem_config.max_page_fault_before = value;
+	}
+	write_unlock(&xpmem_config.lock);
+	return ret;
+}
```

Comment:
can use common function with `xpmem_config_max_page_fault_after_write` to reduce code duplication

Reply:
maybe worth moving all these to debugfs?

Reply:
we agreed to use procfs a the 1st stage
@yosefe

Reply:
merged the code in one check function, still using procfs for now